File: Gestion de la Informacion/Practicas finalizadas/Practica3/src/Laboratorio.java
import java.util.*;

public class Laboratorio 
{
    private static String BD_SERVER = "localhost";
    private static String BD_NAME = "GI1920";

    private String cif;
    private String nombre;
    private String poblacion;
    private Provincia provincia;

    public static List<Laboratorio> ListaLaboratorios()
    {
    	 
        List<Laboratorio> lista = new ArrayList<Laboratorio>();
        // COMPLETAR    
        BD bd = new BD(BD_SERVER, BD_NAME);
        List<Object[]> labos = bd.Select("SELECT cif, nombre, poblacion, provincia from tLaboratorio");
        bd.finalize();
        List<Laboratorio> listaLaboratorios = new ArrayList<>();
        
        for(Object[] row : labos) {
        	String cif = (String) row[0];
        	String nombre = (String) row[1];
        	String poblacion = (String) row[2];
        	Provincia provincia = new Provincia((String)row[3]);
        	
        	Laboratorio laboratorio = new Laboratorio(cif, nombre, poblacion, provincia);
        	listaLaboratorios.add(laboratorio);
        }
        
        return listaLaboratorios;
    }

    public Laboratorio(String cif)
    {
        // COMPLETAR
    	BD bd = new BD(BD_SERVER, BD_NAME);
    	List<Object[]> lab = bd.Select("SELECT nombre, poblacion, provincia from tlaboratorio l where l.cif ='" + cif + "';");
    	bd.finalize();
    	this.cif = cif;
    	this.nombre= lab.get(0).toString();
    	this.poblacion = lab.get(1).toString();
    	this.provincia = new Provincia(lab.get(2).toString());
    }

    public Laboratorio(String cif, String nombre, String poblacion, Provincia provincia)
    {
        // COMPLETAR
    	BD bd = new BD(BD_SERVER, BD_NAME);
    	List<Object[]> lab = bd.Select("SELECT nombre from tlaboratorio l where l.cif ='" + cif + "';");
    	if(lab.isEmpty()) {
        	bd.Insert("INSERT INTO tlaboratorio values ('"+cif+"','"+nombre+"','"+poblacion+", '"+provincia+";)");
    	}
    	bd.finalize();
    	this.cif = cif;
    	this.nombre = nombre;
    	this.poblacion = poblacion;
    	this.provincia = provincia;
    }

    public String getCIF()
    {
    	return cif; 
    }
    
    public void setCIF(String value)       
    {
        // COMPLETAR
    	this.cif = value;
    	BD bd = new BD(BD_SERVER, BD_NAME);
    	bd.Update("UPDATE tlaboratorio set cif='"+this.cif+"';");
    	bd.finalize();
    }

    public String getNombre()
    {
    	return nombre; 
    }
         
    public void setNombre(String value)
    {
        // COMPLETAR
    	this.nombre = value;
    	BD bd = new BD(BD_SERVER, BD_NAME);
    	bd.Update("UPDATE tlaboratorio set nombre='"+this.nombre+"';");
    	bd.finalize();
    }

    public String getPoblacion()
    {
    	return poblacion; 
    }

    public void setPoblacion(String value)
    {
        // COMPLETAR
    	this.poblacion = value;
    	BD bd = new BD(BD_SERVER, BD_NAME);
    	bd.Update("UPDATE tlaboratorio set poblacion='"+this.poblacion+"';");
    	bd.finalize();
    }

    public Provincia getProvincia()
    {
        return provincia;
    }
        
    public void setProvincia (Provincia value)      
    {
        // COMPLETAR
    	this.provincia = value;
    	BD bd = new BD(BD_SERVER, BD_NAME);
    	bd.Update("UPDATE tlaboratorio set provincia='"+this.provincia+"';");
    	bd.finalize();
    	
    }
    
    public void BorrarLaboratorio()
    {
        // COMPLETAR
    	BD bd = new BD(BD_SERVER, BD_NAME);
    	bd.Delete("DELETE FROM tlaboratorio WHERE cif='"+this.cif+"');");
    	bd.finalize();
    	this.cif = null;
    	
    }

    public String toString()
    {
        return cif + " " + nombre;
    }

    public boolean equals(Object obj)
    {
        return obj instanceof Laboratorio 
        		&& (((Laboratorio)obj).cif.compareToIgnoreCase(this.cif) == 0);
    }

    public int hashCode()
    {
        return cif.hashCode();
    }
}

File: Gestion de la Informacion/alumnosP7/Laboratorio.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using BDLibrary;

namespace farmacia
{
    class Laboratorio
    {        
        private string cif;
        private string nombre;
        private string poblacion;
        private Provincia provincia;

        private static string BD_SERVER = Properties.Settings.Default.BD_SERVER;
        private static string BD_NAME = Properties.Settings.Default.BD_NAME;

        public static List<Laboratorio> ListaLaboratorios()
        {
            List<Laboratorio> lista = new List<Laboratorio>();
            SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);
            foreach (object[] tupla in miBD.Select("SELECT CIF FROM tLaboratorio;"))
            {
                string cif = (string)tupla[0];
                Laboratorio l = new Laboratorio(cif);
                lista.Add(l);
            }

            return lista;
        }

        public Laboratorio(string cif)
        {
            SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);
            object[] tupla = miBD.Select("SELECT * FROM tLaboratorio WHERE CIF='" + cif + "';")[0];

            this.cif = (string)tupla[0];
            this.nombre = (string)tupla[1];
            this.poblacion = (string)tupla[2];
            this.provincia = new Provincia((string)tupla[3]);
        }

        public Laboratorio(string cif, string nombre, string poblacion, Provincia provincia)
        {
            cif = cif.ToUpper(); // Siempre en Mayúsculas
            SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);            
            miBD.Insert("INSERT INTO tLaboratorio VALUES('" + cif + "', '" + nombre + "', '" + poblacion + "', '" + provincia.Codigo + "');");
            this.cif = cif;
            this.nombre = nombre;
            this.poblacion = poblacion;
            this.provincia = provincia;
        }

        public string CIF
        {
            get { return cif; }
            set
            {
                SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);
                miBD.Update("UPDATE tLaboratorio SET CIF = '" + value
                    + "' WHERE CIF='" + this.cif + "';");
                cif = value;
            }
        }

        public string Nombre
        {
            get { return nombre; }
            set
            {
                SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);
                miBD.Update("UPDATE tLaboratorio SET Nombre = '" + value
                    + "' WHERE CIF='" + this.cif + "';");
                nombre = value;
            }
        }

        public string Poblacion
        {
            get { return poblacion; }
            set
            {
                SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);
                miBD.Update("UPDATE tLaboratorio SET Poblacion = '" + value
                    + "' WHERE CIF='" + this.cif + "';");
                poblacion = value;
            }
        }

        public Provincia Provincia
        {
            get { return provincia; }
            set
            {
                SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);
                miBD.Update("UPDATE tLaboratorio SET Provincia = '" + value.Codigo
                    + "' WHERE CIF='" + this.cif + "';");
                provincia = value;
            }
        }
        public void BorrarLaboratorio()
        {
            SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);
            miBD.Delete("DELETE FROM tLaboratorio WHERE CIF='" + this.cif + "';");

            this.cif = null;
            this.nombre = null;
            this.poblacion = null;
            this.provincia = null;
        }

        public override string ToString()
        {
            return cif + " " + nombre;
        }

        public override bool Equals(object obj)
        {
            return obj is Laboratorio && (((Laboratorio)obj).CIF.CompareTo(this.CIF) == 0);
        }

        public override int GetHashCode()
        {
            return CIF.GetHashCode();
        }
    }
}
